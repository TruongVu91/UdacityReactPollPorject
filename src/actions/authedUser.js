export const SET_AUTHED_USER = "SET_AUTHED_USER";
export const LOGOUT_USER = "LOGOUT_USER";

export function setAuthedUser(authedUser) {
  return {
    type: SET_AUTHED_USER,
    authedUser,
  };
}

export function logoutUser() {
  return {
    type: LOGOUT_USER,
  };
}

export function handleLogin(username, password) {
  return (dispatch, getState) => {
    const { users } = getState();
    const user = Object.values(users).find(
      (user) => user.id === username && user.password === password
    );
    if (!!user) {
      return dispatch(setAuthedUser(user));
    }
  };
}

export function handleLogout() {
  return (dispatch) => {
    dispatch(logoutUser());
  };
}
