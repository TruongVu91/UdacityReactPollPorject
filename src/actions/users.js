export const RECEIVE_USERS = "RECEIVE_USERS";
export const ADD_USER_QUESTION = "ADD_USER_QUESTION";
export const ADD_ANSWER_USER = "ADD_ANSWER_USER";

export function receiveUsers(users) {
  return {
    type: RECEIVE_USERS,
    users,
  };
}

export function addUserQuestion({ author, id }) {
  return {
    type: ADD_USER_QUESTION,
    author,
    id,
  };
}

export function addAnswerUser(authedUser, qid, answer) {
  return {
    type: ADD_ANSWER_USER,
    authedUser,
    qid,
    answer,
  };
}
