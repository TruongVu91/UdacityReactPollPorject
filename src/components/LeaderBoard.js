import { connect } from "react-redux";
import Table from "react-bootstrap/Table";
import Image from "react-bootstrap/Image";

const LeaderBoard = ({ users }) => {
  return (
    <div>
      <h2 className="h-body">LeaderBoard</h2>
      <Table striped bordered hover>
        <thead>
          <tr>
            <th>Users</th>
            <th>Answered</th>
            <th>Created</th>
          </tr>
        </thead>
        <tbody>
          {users.map((user) => (
            <tr key={user.id}>
              <td>
                <div>
                  <Image
                    className="img-table"
                    src={user.avatarURL}
                    roundedCircle
                  />
                  <br></br>
                  {user.name}
                </div>
              </td>
              <td>{Object.keys(user.answers).length}</td>
              <td>{Object.keys(user.questions).length}</td>
            </tr>
          ))}
        </tbody>
      </Table>
    </div>
  );
};

const mapStateToProps = ({ users }) => ({
  users: Object.values(users).sort(
    (a, b) =>
      Object.keys(b.answers).length +
      Object.keys(b.questions).length -
      (Object.keys(a.answers).length + Object.keys(a.questions).length)
  ),
});

export default connect(mapStateToProps)(LeaderBoard);
