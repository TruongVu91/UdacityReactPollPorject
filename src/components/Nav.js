import Container from "react-bootstrap/Container";
import Nav from "react-bootstrap/Nav";
import Navbar from "react-bootstrap/Navbar";
import Button from "react-bootstrap/Button";
import { connect } from "react-redux";
import Image from "react-bootstrap/Image";
import { handleLogout } from "../actions/authedUser";
import { Link } from "react-router-dom";

const NavBar = ({ dispatch, authedUser }) => {
  const handleLogoutUser = (e) => {
    e.preventDefault();
    console.log("logout");
    dispatch(handleLogout());
  };

  return (
    <Navbar expand="lg" className="bg-body-tertiary">
      <Container>
        <Navbar.Brand>{authedUser ? <h2>Employee Poll</h2> : ""}</Navbar.Brand>
        <Navbar.Toggle aria-controls="basic-navbar-nav" />
        {authedUser ? (
          <Navbar.Collapse id="basic-navbar-nav">
            <Nav className="me-auto">
              <Nav.Link as={Link} to="/">
                Home
              </Nav.Link>
              <Nav.Link as={Link} to="/new" data-testid="newNavLink">
                New
              </Nav.Link>
              <Nav.Link
                as={Link}
                to="/leader-board"
                data-testid="leaderBoardLink"
              >
                LeaderBoard
              </Nav.Link>
            </Nav>
          </Navbar.Collapse>
        ) : (
          ""
        )}
        <Navbar.Brand>
          {authedUser ? (
            <Image
              className="img-table"
              src={authedUser.avatarURL}
              roundedCircle
            />
          ) : (
            ""
          )}
          {authedUser ? <span className="ms-2">{authedUser.name}</span> : ""}
          {authedUser ? (
            <Button
              className="ms-2"
              variant="secondary"
              size="sm"
              onClick={handleLogoutUser}
            >
              Logout
            </Button>
          ) : (
            ""
          )}
        </Navbar.Brand>
      </Container>
    </Navbar>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  authedUser,
});

export default connect(mapStateToProps)(NavBar);
