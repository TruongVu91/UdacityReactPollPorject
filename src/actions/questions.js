import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { addAnswerUser, addUserQuestion } from "./users";

export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const ADD_QUESTION = "ADD_QUESTION";
export const ADD_ANSWER_QUESTION = "ADD_ANSWER_QUESTION";

export function receiveQuestions(questions) {
  return {
    type: RECEIVE_QUESTIONS,
    questions,
  };
}

export function addQuestion(question) {
  return {
    type: ADD_QUESTION,
    question,
  };
}

function addAnswerQuestion(author, qid, answer) {
  return {
    type: ADD_ANSWER_QUESTION,
    author,
    qid,
    answer,
  };
}

export function handleAddQuestion(optionOneText, optionTwoText) {
  console.log("optionOneText", optionOneText);
  console.log("optionTwoText", optionTwoText);
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestion({
      optionOneText,
      optionTwoText,
      author: authedUser,
    }).then((question) => {
      dispatch(addQuestion(question));
      dispatch(addUserQuestion(question));
    });
  };
}

export function handleAddAnswer(questionId, answer) {
  return (dispatch, getState) => {
    const { authedUser } = getState();
    return saveQuestionAnswer({
      authedUser: authedUser.id,
      qid: questionId,
      answer,
    }).then(() => {
      dispatch(addAnswerQuestion(authedUser.id, questionId, answer));
      dispatch(addAnswerUser(authedUser.id, questionId, answer));
    });
  };
}
