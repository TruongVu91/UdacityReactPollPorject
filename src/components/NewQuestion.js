import { useState } from "react";
import { connect } from "react-redux";
import { handleAddQuestion } from "../actions/questions";
import { useNavigate } from "react-router-dom";
import Form from "react-bootstrap/Form";
import Button from "react-bootstrap/Button";

const NewQuestion = ({ dispatch }) => {
  const navigate = useNavigate();
  const [optionOneText, setOptionOneText] = useState("");
  const [optionTwoText, setOptionTwoText] = useState("");

  const handleChangeOptionOneText = (e) => {
    const text = e.target.value;
    setOptionOneText(text);
  };

  const handleChangeOptionTwoText = (e) => {
    const text = e.target.value;
    setOptionTwoText(text);
  };

  const handleSubmitQuestion = (e) => {
    e.preventDefault();
    dispatch(handleAddQuestion(optionOneText, optionTwoText));

    setOptionOneText("");
    setOptionTwoText("");
    navigate("/");
  };

  return (
    <div>
      <h2 className="h-body">Create your own poll</h2>
      <h3 className="center">Would you rather</h3>
      <Form onSubmit={handleSubmitQuestion}>
        <Form.Group className="mb-3">
          <Form.Label>First Option</Form.Label>
          <Form.Control
            type="input"
            placeholder="Enter first option"
            value={optionOneText}
            onChange={handleChangeOptionOneText}
          />
        </Form.Group>
        <Form.Group className="mb-3">
          <Form.Label>Second Option</Form.Label>
          <Form.Control
            type="input"
            placeholder="Enter second option"
            value={optionTwoText}
            onChange={handleChangeOptionTwoText}
          />
        </Form.Group>
        <Button
          variant="primary"
          type="submit"
          disabled={optionOneText === "" || optionTwoText === ""}
        >
          Submit
        </Button>
      </Form>
    </div>
  );
};

export default connect()(NewQuestion);
