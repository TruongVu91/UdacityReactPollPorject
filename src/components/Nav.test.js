/* eslint-disable testing-library/prefer-screen-queries */
/* eslint-disable testing-library/render-result-naming-convention */
import "@testing-library/jest-dom/extend-expect";

import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { createStore } from "redux";
import middleware from "../middleware";
import reducers from "../reducers";
import { BrowserRouter } from "react-router-dom";
import { setAuthedUser } from "../actions/authedUser";
import Nav from "./Nav";
import NewQuestion from "./NewQuestion";
import LeaderBoard from "./LeaderBoard";

const store = createStore(reducers, middleware);

describe("Nav", () => {
  it("check display nav bar", () => {
    store.dispatch(setAuthedUser({ id: "sarahedo", password: "" }));

    const navView = render(
      <Provider store={store}>
        <BrowserRouter>
          <Nav />
        </BrowserRouter>
      </Provider>
    );
    expect(navView).toBeDefined();
    expect(navView).toMatchSnapshot();
  });

  it("check display new question view", () => {
    store.dispatch(setAuthedUser({ id: "sarahedo", password: "" }));

    const newQuestionView = render(
      <Provider store={store}>
        <BrowserRouter>
          <NewQuestion />
        </BrowserRouter>
      </Provider>
    );

    expect(newQuestionView).toBeDefined();
    expect(newQuestionView).toMatchSnapshot();
  });

  it("check display LeaderBoard view", () => {
    store.dispatch(setAuthedUser({ id: "sarahedo", password: "" }));

    const leaderBoardView = render(
      <Provider store={store}>
        <BrowserRouter>
          <LeaderBoard />
        </BrowserRouter>
      </Provider>
    );

    expect(leaderBoardView).toBeDefined();
    expect(leaderBoardView).toMatchSnapshot();
  });
});
