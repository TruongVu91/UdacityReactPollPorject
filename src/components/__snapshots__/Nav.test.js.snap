// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav check display LeaderBoard view 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2
          class="h-body"
        >
          LeaderBoard
        </h2>
        <table
          class="table table-striped table-bordered table-hover"
        >
          <thead>
            <tr>
              <th>
                Users
              </th>
              <th>
                Answered
              </th>
              <th>
                Created
              </th>
            </tr>
          </thead>
          <tbody />
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2
        class="h-body"
      >
        LeaderBoard
      </h2>
      <table
        class="table table-striped table-bordered table-hover"
      >
        <thead>
          <tr>
            <th>
              Users
            </th>
            <th>
              Answered
            </th>
            <th>
              Created
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Nav check display nav bar 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        class="bg-body-tertiary navbar navbar-expand-lg navbar-light"
      >
        <div
          class="container"
        >
          <span
            class="navbar-brand"
          >
            <h2>
              Employee Poll
            </h2>
          </span>
          <button
            aria-controls="basic-navbar-nav"
            aria-label="Toggle navigation"
            class="navbar-toggler collapsed"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="navbar-collapse collapse"
            id="basic-navbar-nav"
          >
            <div
              class="me-auto navbar-nav"
            >
              <a
                class="nav-link"
                href="/"
              >
                Home
              </a>
              <a
                class="nav-link"
                data-testid="newNavLink"
                href="/new"
              >
                New
              </a>
              <a
                class="nav-link"
                data-testid="leaderBoardLink"
                href="/leader-board"
              >
                LeaderBoard
              </a>
            </div>
          </div>
          <span
            class="navbar-brand"
          >
            <img
              class="img-table rounded-circle"
            />
            <span
              class="ms-2"
            />
            <button
              class="ms-2 btn btn-secondary btn-sm"
              type="button"
            >
              Logout
            </button>
          </span>
        </div>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      class="bg-body-tertiary navbar navbar-expand-lg navbar-light"
    >
      <div
        class="container"
      >
        <span
          class="navbar-brand"
        >
          <h2>
            Employee Poll
          </h2>
        </span>
        <button
          aria-controls="basic-navbar-nav"
          aria-label="Toggle navigation"
          class="navbar-toggler collapsed"
          type="button"
        >
          <span
            class="navbar-toggler-icon"
          />
        </button>
        <div
          class="navbar-collapse collapse"
          id="basic-navbar-nav"
        >
          <div
            class="me-auto navbar-nav"
          >
            <a
              class="nav-link"
              href="/"
            >
              Home
            </a>
            <a
              class="nav-link"
              data-testid="newNavLink"
              href="/new"
            >
              New
            </a>
            <a
              class="nav-link"
              data-testid="leaderBoardLink"
              href="/leader-board"
            >
              LeaderBoard
            </a>
          </div>
        </div>
        <span
          class="navbar-brand"
        >
          <img
            class="img-table rounded-circle"
          />
          <span
            class="ms-2"
          />
          <button
            class="ms-2 btn btn-secondary btn-sm"
            type="button"
          >
            Logout
          </button>
        </span>
      </div>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Nav check display new question view 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <h2
          class="h-body"
        >
          Create your own poll
        </h2>
        <h3
          class="center"
        >
          Would you rather
        </h3>
        <form
          class=""
        >
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              First Option
            </label>
            <input
              class="form-control"
              placeholder="Enter first option"
              type="input"
              value=""
            />
          </div>
          <div
            class="mb-3"
          >
            <label
              class="form-label"
            >
              Second Option
            </label>
            <input
              class="form-control"
              placeholder="Enter second option"
              type="input"
              value=""
            />
          </div>
          <button
            class="btn btn-primary"
            disabled=""
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <h2
        class="h-body"
      >
        Create your own poll
      </h2>
      <h3
        class="center"
      >
        Would you rather
      </h3>
      <form
        class=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            First Option
          </label>
          <input
            class="form-control"
            placeholder="Enter first option"
            type="input"
            value=""
          />
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
          >
            Second Option
          </label>
          <input
            class="form-control"
            placeholder="Enter second option"
            type="input"
            value=""
          />
        </div>
        <button
          class="btn btn-primary"
          disabled=""
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
